tests/test_responses.py

```diff
@@ -0,0 +1,122 @@
+from base64 import b64encode
+
+import pytest
+
+from scrapy_zyte_api.responses import ZyteAPIResponse, ZyteAPITextResponse
+
+PAGE_CONTENT = "<html><body>The cake is a lie!</body></html>"
+URL = "https://example.com"
+
+
+def api_response_browser():
+    return {
+        "url": URL,
+        "browserHtml": PAGE_CONTENT,
+        "javascript": True,
+        "echoData": {"some_value": "here"},
+        "httpResponseHeaders": [
+            {"name": "Content-Type", "value": "text/html"},
+            {"name": "Content-Length", "value": len(PAGE_CONTENT)},
+        ],
+    }
+
+
+def api_response_body():
+    return {
+        "url": "https://example.com",
+        "httpResponseBody": b64encode(PAGE_CONTENT.encode("utf-8")),
+        "echoData": {"some_value": "here"},
+        "httpResponseHeaders": [
+            {"name": "Content-Type", "value": "text/html"},
+            {"name": "Content-Length", "value": len(PAGE_CONTENT)},
+        ],
+    }
+
+
+EXPECTED_HEADERS = {b"Content-Type": [b"text/html"], b"Content-Length": [b"44"]}
+EXPECTED_BODY = PAGE_CONTENT.encode("utf-8")
+
+
+@pytest.mark.parametrize(
+    "api_response,cls",
+    [
+        (api_response_browser, ZyteAPITextResponse),
+        (api_response_body, ZyteAPIResponse),
+    ],
+)
+def test_init(api_response, cls):
+    response = cls(URL, zyte_api_response=api_response())
+    assert response.zyte_api_response == api_response()
+
+    assert response.url == URL
+    assert response.status == 200
+    assert not response.headers
+    assert response.body == b""
+    assert not response.flags
+    assert response.request is None
+    assert response.certificate is None
+    assert response.ip_address is None
+    assert response.protocol is None
+
+
+@pytest.mark.parametrize(
+    "api_response,cls",
+    [
+        (api_response_browser, ZyteAPITextResponse),
+        (api_response_body, ZyteAPIResponse),
+    ],
+)
+def test_text_from_api_response(api_response, cls):
+    response = cls.from_api_response(api_response())
+    assert response.zyte_api_response == api_response()
+
+    assert response.url == URL
+    assert response.status == 200
+    assert response.headers == EXPECTED_HEADERS
+    assert response.body == EXPECTED_BODY
+    assert response.flags == ["zyte-api"]
+    assert response.request is None
+    assert response.certificate is None
+    assert response.ip_address is None
+    assert response.protocol is None
+
+
+@pytest.mark.parametrize(
+    "api_response,cls",
+    [
+        (api_response_browser, ZyteAPITextResponse),
+        (api_response_body, ZyteAPIResponse),
+    ],
+)
+def test_response_replace(api_response, cls):
+    orig_response = cls.from_api_response(api_response())
+
+    # The ``zyte_api_response`` should not be replaced.
```

Comment:
Hm, why shouldn't it be possible to replace it? 
If that's a feature which is desirable, but not implemented, we could have an xfail test for this.

Reply:
Hi @kmike , this was originally intentional as to prevent the original Zyte API response from being unintentionally replaced. However, I just realized that even other attributes like `url` and `body` could also be replaced. Thus, `zyte_api_response` should have the same expectation like them.

Updated this with an **xfail** in https://github.com/scrapy-plugins/scrapy-zyte-api/pull/10/commits/84dac7de138f50f6e0727384fafccad6062817a5 as I think the ability to preserve the original Zyte API responses is desirable later on. Kindly let me know otherwise.

Reply:
If `zyte_api` is passed to `replace`, then the intention of the coder is to replace it, and I think it should be replaced. If not passed (existing middlewares will not pass it, only new third-party or user ones may), it should stay with the same value.

I wonder if we should turn the property into a regular `zyte_api`-named public attribute, and implement this `replace` method in line with pre-2.6 `TextResponse`:

```python
    def replace(self, *args, **kwargs):
        kwargs.setdefault('zyte_api', self.zyte_api)
        return super().replace(*args, **kwargs)
```

[2.6 introduced Response.attributes](https://github.com/scrapy/scrapy/pull/5218), which I guess we cannot use here since we want support for older Scrapy.

Reply:
Makes sense. I think we can use the new `Response.attributes` in **2.6** since a lot of the libs we're working on sets it as the minimum version. Updated in https://github.com/scrapy-plugins/scrapy-zyte-api/pull/10/commits/5b7cf6fded737bb08c0d03b422c50120f920e2ed.

For example, https://github.com/scrapinghub/scrapy-poet/pull/62 needs **2.6** for the non-experimental `asyncio` support.